File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.py
import sys
input = sys.stdin.readline

def solution(N, r, c):
    global white, blue
    color = A[r][c]
    for i in range(r, r+N):
        for j in range(c, c+N):
            if color != A[i][j]:
                solution(N//2, r, c)
                solution(N//2, r+N//2, c)
                solution(N//2, r, c+N//2)
                solution(N//2, r+N//2, c+N//2)
                return
    if color == 0:
        white += 1
    else:
        blue += 1

N = int(input())
A = [list(map(int, input().split())) for _ in range(N)]
#print(A)

white = 0
blue = 0
solution(N, 0, 0)#색종이 크기, r, c

print(white)
print(blue)

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;


public class Main {
    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static int N;
    private static int[][] map;
    private static int white;
    private static int blue;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());
        map = new int[N][N];

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        solution(0, 0, N);
        System.out.println(white);
        System.out.println(blue);
    }

    private static void solution(int r, int c, int N) {
        int tmp = map[r][c];
        for (int i = r; i < r+N; i++) {
            for (int j = c; j < c+N; j++) {
                if (tmp != map[i][j]) {
                    solution(r, c, N / 2);
                    solution(r, c + N / 2, N / 2);
                    solution(r + N / 2, c, N / 2);
                    solution(r + N / 2, c + N / 2, N / 2);
                    return;
                }
            }
        }
        if (tmp == 0) {
            white += 1;
        } else if (tmp == 1) {
            blue += 1;
        }
    }

}